File: Jim/src/sk/fiit/jim/garbage/plan/PlanTournamentKickDistance.java
package sk.fiit.jim.garbage.plan;


import sk.fiit.jim.LambdaCallable;
import sk.fiit.jim.agent.AgentInfo;
import sk.fiit.jim.agent.highskill.Beam;
import sk.fiit.jim.agent.highskill.GetUp;
import sk.fiit.jim.agent.highskill.KickTournament;
import sk.fiit.jim.agent.highskill.Localize;
import sk.fiit.jim.agent.highskill.TurnToPosition;
import sk.fiit.jim.agent.highskill.Walk2BallTournament;
import sk.fiit.jim.agent.models.AgentModel;
import sk.fiit.jim.agent.models.EnvironmentModel;
import sk.fiit.jim.agent.models.WorldModel;
import sk.fiit.robocup.library.geometry.Vector3D;

/**
 * Rewrited PlanTournamentGetUpFromBelly from ruby
 * 
 * @author Matej Badal
 */

public class PlanTournamentKickDistance extends Plan {
	
	public static PlanTournamentKickDistance getInstance() {
		return new PlanTournamentKickDistance();
	}
	
	public void replan() {
		AgentInfo agentInfo = AgentInfo.getInstance();
        AgentModel agentModel = AgentModel.getInstance();
        Vector3D kick_target = agentInfo.kickTarget();
		
		
		if (EnvironmentModel.beamablePlayMode() && !EnvironmentModel.isKickOffLeftPlayMode()) {
			AgentInfo.logState("Beam");
			Vector3D start_position = Vector3D.cartesian(-5, 1, 0.4);
			Highskill_Queue.add(new Beam(start_position));
			this.beamed = true;
		} else if (agentModel.isOnGround() || agentModel.isLyingOnBack()
					|| agentModel.isLyingOnBelly()) {
			AgentInfo.logState("GetUp");
			Highskill_Queue.add(new GetUp());
		} else if (!WorldModel.getInstance().isSeeBall()) {
			AgentInfo.logState("Localize");
			Highskill_Queue.add(new Localize());
		} else if (!this.is_ball_mine() || !this.straight()) {
			AgentInfo.logState("Walk");
			boolean a = !is_ball_mine() && !this.straight();
			boolean b = EnvironmentModel.beamablePlayMode() && !EnvironmentModel.isKickOffLeftPlayMode();
			Highskill_Queue.add(new Walk2BallTournament(a && !b));
		} else if (!this.turned_to_goal()) {
			AgentInfo.logState("Turn");
			Highskill_Queue.add(new TurnToPosition(kick_target,
					new LambdaCallable() {
						public boolean call() {
							return ! turned_to_goal();
						}
					}
					));
		} else if (this.is_ball_mine() && this.straight() && this.turned_to_goal()) {
			AgentInfo.logState("Kick");
			Highskill_Queue.add(new KickTournament(kick_target));
		} else {
			AgentInfo.logState("???");
			Highskill_Queue.add(new GetUp());
		}
	}
}

File: Old_xml_and_scripts/Jim/plan/29.5.2013/planTournamentKickDistance.rb
### planTournamentKickDistance.rb ###
#	@author: A55Kickers				#
#####################################

class PlanTournamentKickDistance < Plan
  
 
  def replan
    @plan.clear
    @agentInfo.whereIsBall();
    @agentInfo.isBallMine();
    @agentInfo.nearBall();
    
    @agentInfo.ballControlPosition();
	 
	@mathModel = Java::sk.fiit.robocup.library.geometry.Vector3D.clone
    @target_position = @agentInfo.ballControlPosition()
    @target_position_phi = @target_position.getPhi()
    
    # miesto kde ma kupnut
    @kick_target = @mathModel.cartesian(10.5,0.0,0.0) #@agentInfo.kickTarget()
    
    # vypocet uhla pre nastavenie sa na poziciu
    me_flattened = my.position.set_z(0.0)
    target_vector = @kick_target - me_flattened
    target_angle = target_vector.phi
    diff_against_current = Angles.normalize(target_angle - my.rotation_z)    
    #uhol natocenia
    @kick_target_phi = diff_against_current #@kick_target.getPhi()
    
     if EnvironmentModel.beamablePlayMode and not EnvironmentModel.isKickOffLeftPlayMode
    	@agentInfo.loguj('Beam');
    	@plan << Beam.new(Java::sk.fiit.robocup.library.geometry.Vector3D.cartesian(-0.4, 0, 0.4))
    	@beamed = true
      	    
    elsif (me.on_ground? or me.is_lying_on_back? or me.is_lying_on_belly?)
    	@agentInfo.loguj('GetUp');
     	@plan << GetUp.new(Proc.new{me.on_ground?})
    elsif (not see_ball?) 
   		@agentInfo.loguj('Localize');
    	@plan << Localize.new(Proc.new{(not see_ball?)})
    elsif (not is_ball_mine? or not straight?)
    	@agentInfo.loguj('Walk');
    	@plan << Walk2Ball.new(:ball, Proc.new{(not is_ball_mine? or not straight?) and not (EnvironmentModel.beamablePlayMode and not EnvironmentModel.isKickOffLeftPlayMode)})# or (not see_ball?)})
    elsif not turned_to_goal?
    	@agentInfo.loguj('Turn');
     	@plan << TurnToPosition.new(@kick_target, Proc.new{not turned_to_goal?})
    elsif (is_ball_mine? and straight? and turned_to_goal?)
      @agentInfo.loguj('Kick');
    	@plan << Kick.new(@kick_target)
    else
    	@agentInfo.loguj('???');
    	@plan << GetUp.new(Proc.new{me.on_ground?})
    end
  end
end